File: 003_Longest_Substring_Without_Repeating_Characters/003_3.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        length = 0
        left = 0
        last = {}

        for i, char in enumerate(s):
            if char in last and last[char] >= left:
                left = last[char] + 1
            last[char] = i
            length = max(length, i - left + 1)
        return length

File: 003_Longest_Substring_Without_Repeating_Characters/003_3.cpp
class Solution {
public:
	int lengthOfLongestSubstring(string s) {
		int length = 0;
		unordered_map<char, int> map;
		int size = (int)s.size();
		for (int i = 0, j = 0; j < size; ++j) {
			if (map.find(s[j]) != map.end())
				i = max(map[s[j]] + 1, i);
			map[s[j]] = j;
			length = max(length, j - i + 1);
		}
        return length;
	}
};